Implementation for post likes endpoint

bitdc opened the pull request:
### Description

This PR adds an endpoint for fetching post likes based on [this API documentation](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/%24post_ID/likes/). Since the likes are contextual to the Post object, the implementation is added in `PostServiceRemoteREST`. The response is then mapped to the existing `RemoteUser` object.

References:
- Related issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/15662
- Design: [Likes Enhancements i2](https://wp.me/pbArwn-1CQ)

### Testing Details

Added test that ensures the object returned from the API are correctly parsed into `RemoteUser`s, and a negative test that ensures the failure block is called.

- [x] Please check here if your pull request includes additional test coverage.

bitdc committed:
Add method signature for fetching post likes

bitdc committed:
Add implementation for fetching post likes

bitdc committed:
Add tests for fetching post likes

bitdc committed:
Update docs on PostServiceRemoteREST

bitdc committed:
Minor cleanup on tests

We don't need to call fulfill when an XCTFail is already called.

bitdc committed:
Fix closure syntax for CI

bitdc committed:
Revert nullability attributes in PostServiceRemote

This triggers warning on the validate-podspec workflow, since other
methods are not attributed with nullability specifiers.

bitdc committed:
Update test with optional type

bitdc committed:
Move endpoint to PostServiceRemoteREST

ScoutHarris commented on WordPressKit/PostServiceRemoteREST.h:
```diff
@@ -53,4 +55,18 @@
+ *  @param      success     The block that will be executed on success. Can be nil.
+ *  @param      failure     The block that will be executed on failure. Can be nil.
+ */
+- (void)getLikesForID:(NSNumber *)postID
```
Can we make the method name a bit clearer? `ID` is a bit vague. Maybe something like `getLikesforPost` instead?

bitdc replied:
Yeah, sure! Would `getLikesForPostID` be better here? Because there are other methods that require a `RemotePost` type as the first parameter, such as `-getAutoSaveForPost:(RemotePost *)post`.

ScoutHarris replied:
Yep, that'll work too!

ScoutHarris commented on WordPressKit/PostServiceRemoteREST.m:
```diff
@@ -306,6 +307,30 @@ - (void)restorePost:(RemotePost *)post
+                       parameters:nil
+                          success:^(id responseObject, NSHTTPURLResponse *httpResponse) {
+        NSArray *jsonUsers = responseObject[@"likes"];
+        if (success && jsonUsers) {
```
It looks like nothing is returned if `jsonUsers` is not present. Shouldn't we just check for `success` and return an empty array of there are no likes?

bitdc replied:
Nice catch! Will fix this in the next commit.

ScoutHarris commented on WordPressKit/PostServiceRemoteREST.m:
```diff
@@ -564,4 +589,42 @@ - (NSArray *)tagNamesFromJSONDictionary:(NSDictionary *)jsonTags {
+ *
+ *  @param  jsonUser    The dictionary that represents a RemoteUser.
+ */
+- (RemoteUser *)remoteUserFromJSONDictionary:(NSDictionary *)jsonUser
```
From your success test, it looks like `email` is included in the response. Not sure if it is ever actually provided, but just in case, can we set `user.email` if it's present and default to `""` if not?

bitdc replied:
Ah yes, I initially thought that too, but the problem is the `email` parameter returned from the API is a boolean. Here's an example of a complete user JSON from post likes endpoint:
```json
{
    "found": 1,
    "i_like": true,
    "site_ID": 0,
    "post_ID": 1,
    "likes": [
        {
            "ID": 106059135,
            "login": "dvdchr",
            "email": false,
            "name": "dvdchr",
            "first_name": "David",
            "last_name": "Christiandy",
            "nice_name": "dvdchr",
            "URL": "",
            "avatar_URL": "https://0.gravatar.com/avatar/9d5db177ff5c19b2c4806093ec6d87f7?s=96&d=identicon&r=G",
            "profile_URL": "https://en.gravatar.com/dvdchr",
            "ip_address": false,
            "site_ID": 174074163,
            "site_visible": true,
            "default_avatar": false
        }
    ]
}

```
In fact, I don't know what the boolean value in the `email` field represents, as I find that all of the JSON objects have `false` value for every entry. 😅

ScoutHarris replied:
Ah, I see. Yea, that's a little odd. 🤷 OK. We'll leave it for now. I'm not sure if we'll ever need it in this scenario, but if so we can come back to it. Thanks for checking!

ScoutHarris commented:
Hey David. Just a few minor questions/suggestions. Otherwise, this is looking good! 🎉 

(Sorry, I forgot to comment when I submitted my review.)

bitdc committed:
Call success block when response is empty

Referring to this review:
https://github.com/wordpress-mobile/WordPressKit-iOS/pull/347#discussion_r572372116,
in the Post fetch likes method, ensure that success block is called when
the response is empty.

bitdc committed:
Update method name to getLikesForPostID

References to this review: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/347#discussion_r572370871

ScoutHarris reviewed (approved):
Looks good! Nice job!
